src/wallet/scriptpubkeyman.cpp

```diff
@@ -1717,42 +1717,160 @@ std::unordered_set<CScript, SaltedSipHasher> LegacyDataSPKM::GetScriptPubKeys()
-                    spks.insert(ms_spk);
+            break;
+        }
+        case TxoutType::MULTISIG:
+        {
+            // Multisigs are only spendable if we have all of their keys
+            std::vector<std::vector<unsigned char>> keys(sols.begin() + 1, sols.begin() + sols.size() - 1);
+            if (!HaveKeys(keys, *this)) {
+                break;
+            }
+            // Multisigs are always spendable inside of P2SH scripts
+            spks.insert(GetScriptForDestination(ScriptHash(script)));
+            // We need to have the P2WSH script for the P2WSH to be spendable.
+            // But the keys also must be compressed
+            CScript ms_wsh = GetScriptForDestination(WitnessV0ScriptHash(script));
+            if (HaveCScript(CScriptID(ms_wsh))) {
+                if (all_keys_compressed(keys)) {
+                    spks.insert(ms_wsh);
+                    spks.insert(GetScriptForDestination(ScriptHash(ms_wsh)));
                 }
             }
+            break;
+        }
+        }
+    }
+    // Iterate again for all the P2WSH scripts
+    for (const auto& script_pair : mapScripts) {
+        const CScript& script = script_pair.second;
+        std::vector<std::vector<unsigned char>> sols;
+        TxoutType type = Solver(script, sols);
+        if (type == TxoutType::WITNESS_V0_SCRIPTHASH) {
+            uint160 hash{RIPEMD160(sols[0])};
+            CScript witness_script;
+            int wit_ver = -1;
+            std::vector<unsigned char> wit_prog;
+            if (GetCScript(CScriptID(hash), witness_script) &&
+                !witness_script.IsPayToScriptHash() &&
+                !witness_script.IsWitnessProgram(wit_ver, wit_prog) &&
+                spks.count(witness_script) > 0) {
+                spks.insert(script);
+                spks.insert(GetScriptForDestination(ScriptHash(script)));
+            }
         }
     }
 
+    enum class ScriptContext {
+        TOP,
+        P2SH,
+        P2WSH,
+    };
+    std::function<bool(const CScript&, const ScriptContext)> is_valid_script = [&](const CScript& script, const ScriptContext ctx) -> bool {
+        std::vector<valtype> sols;
+        TxoutType spk_type = Solver(script, sols);
+
+        CKeyID keyID;
+        switch (spk_type) {
+        case TxoutType::NONSTANDARD:
+        case TxoutType::NULL_DATA:
+        case TxoutType::WITNESS_UNKNOWN:
+        case TxoutType::WITNESS_V1_TAPROOT:
+        case TxoutType::ANCHOR:
+            return ctx == ScriptContext::TOP;
+        case TxoutType::PUBKEY:
+            if (ctx == ScriptContext::P2WSH && sols[0].size() != 33) return false;
+            return true;
+        case TxoutType::WITNESS_V0_KEYHASH:
+            return ctx != ScriptContext::P2WSH;
+        case TxoutType::PUBKEYHASH:
+            if (ctx == ScriptContext::P2WSH) {
+                CPubKey pubkey;
+                if (GetPubKey(CKeyID(uint160(sols[0])), pubkey) && !pubkey.IsCompressed()) {
```

Comment:
What is our expectation to be happening when `uint160(sols[0])` is called here? Could we maybe pull apart these three conversations in a conditional statement into two lines where the first line provides some insight on what we are converting here?

Would I be right in guessing that a `CKeyID` is the hash of a public key?

Reply:
It's basically just casting types.

`sols[0]` contains a vector of bytes which represents the hash of a pubkey. This is cast to the `uint160` type, which is then cast to the `CKeyID` type (which is really just a wrapper around `uint160`) which is the type that `GetPubKey` can take. There is no direct conversion from `std::vector<>` to `CKeyID` hence the `uint160` in between.

It doesn't make sense to pull this apart since the intermediate types are never used.